### ISSUE TITLE
[Feature] VNester: omit aliases

### ISSUE THREAD

**Issue by Poweranimal** [lulupointu/vrouter#127]
Hi @lulupointu ,

I'm very grateful for your `vrouter` package that you maintain.
I really like working with it because of its straight forward api.

Speaking about such, the only component that doesn't feel straight forward to use is the `VNester`.
In that regard, it's the required use of `aliases` for common behaviors.
The following points are derived from this issue, including its code examples: [
[Question] VNester: stacked route on all nested pages](https://github.com/lulupointu/vrouter/issues/126#issue-958208595)

Issues I have with using aliases:
1. An `alias` should only do, what it describes to be: being an alias path.
2. Using `VNester` doesn't make it obvious that for common behaviors an `alias` is needed. One needs to carefully read the doc at [Custom VRouteElement And Scaling](https://vrouter.dev/guide/Advanced/Custom%20VRouteElement%20And%20Scaling).
3. In more complex setups with `VNester` one needs to add the same alias to many different places. This seems to be error prune and isn't a nice usage experience.

I'm not an expert at all in flutter or especially in routing. However, I would suggest if it's technically possible, to get rid of the use of aliases for `VNester`.

Doing so would also make the usage of the `VNester` component as straight forward as any other components of this great package. 

I look forward to your response.

**Comment by Poweranimal** [comment]
The following is an example how I could image it to look like.
The example is derived from [#126](https://github.com/lulupointu/vrouter/issues/126#issue-958208595)

```dart
class ProfileRoute extends VRouteElementBuilder {
  static final String profile = '/profile';
  static final String settings = profile + '/settings';

  @override
  List<VRouteElement> buildRoutes() {
    return [
      ScaffoldRouteElement(
        path: profile,
        nestedRoutes: [
          VWidget(
            key: animationKey,
            path: null,
            aliases: ['buy'],
            widget: ContentScreen(
              title: 'Profile',
              actions: [
                ContentScreenAction(
                  title: 'Show Settings',
                  action: (context) => context.vRouter.to(settings),
                ),
              ],
            ),
            stackedRoutes: [
              VWidget(
                path: settings,
                // Maybe also getting rid of aliases here?
                aliases: [settings + '/buy'],
                key: ValueKey('Order'),
                widget: ContentScreen(title: 'Settings'),
              ),
            ],
          ),
        ],
        stackedRoutes: [
          VWidget(
            path: 'buy',
            widget: ContentScreen(title: 'Buy'),
            fullscreenDialog: true,
          ),
          VWidget(
            path: '*/buy',
            widget: ContentScreen(title: 'Buy'),
            fullscreenDialog: true,
          ),
        ],
      ),
    ];
  }

```

**Comment by lulupointu** [comment]
Exactly, you could event have 1 path with `*buy`.

I understand the trouble but the issue is that some people use an absolute path on the stack route. I don't see how to make it easier while still allowing both scenarios. I believe using `*buy` is actually a great solution for relative paths!

**Comment by Poweranimal** [comment]
I can't follow the issue regarding absolute paths.
Would you be so kind and give an example?

**Comment by lulupointu** [comment]
It's not so much an issue as much as a different use case which makes simplification impossible (as far as I see).

### Relative paths in `VNester.stackedRoutes`

<details>
<summary>
If we only had relative path in `VNester.stackedRoutes` we could have:
</summary>

Note that bellow is not working code, it's just an example of what could be achieved if we assumed that only relative paths where used in `VNester.stackedRoutes`.

```dart
ScaffoldRouteElement(
  path: null, // CHANGE: Path moved down where it belongs
  nestedRoutes: [
    VWidget(
      path: profile, // CHANGE: Path moved here, no alias (which also implies not even needing a key)
      widget: ContentScreen(
        title: 'Profile',
        actions: [
          ContentScreenAction(
            title: 'Show Settings',
            action: (context) => context.vRouter.to(settings),
          ),
        ],
      ),
      stackedRoutes: [
        VWidget(
          path: settings, // CHANGE: No alias here (which also implies not even needing a key)
          widget: ContentScreen(title: 'Settings'),
        ),
      ],
    ),
  ],
  stackedRoutes: [
    VWidget(
      path: 'buy', // CHANGE: Omit alias
      widget: ContentScreen(title: 'Buy'),
      fullscreenDialog: true,
    ),
  ],
)
```

In this case where the path is relative it's pretty easy for me to guess that:
- `/profile/buy` is a valid path, with `ProfileScreen` bellow `BuyScreen`. And we should pop on `/profile` 
- `/profile/settings/buy` is a valid path, with `SettingsScreen` bellow `BuyScreen`. And we should pop on `/profile/settings`

It's a pretty easy process: to get the path just remove the part of the `stackedRoutes` and get the new path. This is great and would solve every issues you have. However as you will see bellow this does not extend very well on absolute path.

</details>

### Absolute paths in `VNester.stackedRoutes`

<details>
<summary>
If we now have an absolute path in `VNester.stackedRoutes`:
</summary>

```dart
ScaffoldRouteElement(
  path: profile, // CHANGE: we need the path here to know where to pop
  nestedRoutes: [
    VWidget(
      key: animationKey, // CHANGE: We do need the key now because of aliases
      path: null, // CHANGE: we need null to match parent path
      aliases: ['/buy'], // CHANGE: we need alias to tell that this should be show bellow BuyScreen
      widget: ContentScreen(
        title: 'Profile',
        actions: [
          ContentScreenAction(
            title: 'Show Settings',
            action: (context) => context.vRouter.to(settings),
          ),
        ],
      ),
      stackedRoutes: [
        VWidget(
          path: settings,
          key: ValueKey('Order'),
          widget: ContentScreen(title: 'Settings'),
        ),
      ],
    ),
  ],
  stackedRoutes: [
    VWidget(
      path: '/buy', // CHANGE: absolute path
      widget: ContentScreen(title: 'Buy'),
      fullscreenDialog: true,
    ),
  ],
)
```

Now this one is a bit more tricky, it requires the use of 3 unintuitive things:
-  `null` path to match the parent, because the path needs to be moved up to allow pop to work properly
-  an `alias` to know what to show bellow `BuyScreen`
- A key to prevent the animation in the `ProfileScreen`

But more importantly, we can't apply the same logic as before:
- We can't easily deduce where to pop
- We can't easily deduce what to show in `VNester.nestedRoutes`

</details>

I hope you understand what I mean and appreciate how those scenarios differ and force the API to be how it is. I'm sure there is a better way, maybe abstract this logic in some way. If you have any idea please share, otherwise I'm afraid I don't have enough brain power available at the moment to find it.

**Comment by Poweranimal** [comment]
Thank you very much for the explanation.
I'll try my best to work out an implementation draft.

**Comment by marcguilera** [comment]
To allow to navigate from inside of an `VNester` is the suggested approach to add the stacked routes to it AND the inner `VRouteElement`? With this I see the correct animations but it seems like animations/navigations are happening twice when checking the `NavigatorObserver`. 

Example logs:
```
flutter: 👻 11:47:26.163196 INFO     Router - push: previousRoute=null, route=null
flutter: 👻 11:47:26.277901 INFO     Router - push: previousRoute=null, route=/feed
flutter: 👻 11:51:13.794022 INFO     Router - push: previousRoute=null, route=/categories/1
flutter: 👻 11:51:13.826619 INFO     Router - push: previousRoute=/feed, route=/categories/1
flutter: 👻 11:51:15.186098 INFO     Router - pop: previousRoute=/categories/1, route=null
flutter: 👻 11:51:15.191533 INFO     Router - pop: previousRoute=/categories/1, route=/feed
```

If I only add the stack to the inner `VWidget` the page is displayed under the menu, which I expect but is not a what I want. Is there a way to remove this duplication?

In short, the usecase is:
- A root home containing a `VNester` with the bottom nav bar. 
- A couple of pages which should be under it.
- Being able to navigate from one of those pages to another but over everything else (ie: detail page). 

The last point is what I'm trying to understand. Thank you!

**Comment by lulupointu** [comment]
@marcguilera here is a code sample that might enlighten you

<details>
<summary>
 Read the comment to have information on the things that might not be obvious
</summary>

```dart
import 'package:flutter/material.dart';
import 'package:vrouter/vrouter.dart';

void main() {
  runApp(
    VRouter(
      debugShowCheckedModeBanner: false,
      routes: [
        VNester(
          path: '/user', // We use '/user' so that [VNester.stackedRoute] knows on what to pop
          widgetBuilder: (child) => MyScaffold(child), // Child is the widget from nestedRoutes
          nestedRoutes: [
            VWidget(path: '/', widget: HomeScreen()),
            VWidget(
              key: ValueKey('UserScreen'),
              // We use a key to indicate that path and alias lead to the same screen
              path: null,
              // null path matches parent
              aliases: ['/settings'],
              // We use an alias because we need [VNester] to know what to use as a nestedRoute when on '/settings'
              widget: UserScreen(),
            ),
          ],
          stackedRoutes: [
            VWidget(path: '/settings', widget: SettingsScreen()),
          ],
        ),
      ],
    ),
  );
}

class BaseWidget extends StatelessWidget {
  final String title;

  final String buttonText;

  final String to;

  const BaseWidget({
    Key? key,
    required this.title,
    required this.buttonText,
    required this.to,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Material(
      child: Scaffold(
        appBar: AppBar(title: Text(title)),
        body: Center(
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              Text(title),
              SizedBox(height: 50),
              ElevatedButton(
                onPressed: () => context.vRouter.to(to),
                child: Text(buttonText),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class MyScaffold extends StatelessWidget {
  final Widget child;

  const MyScaffold(this.child);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: child,
      bottomNavigationBar: BottomNavigationBar(
        currentIndex: context.vRouter.url.contains('user') ? 1 : 0,
        onTap: (value) => context.vRouter.to((value == 0) ? '/' : '/user'),
        items: [
          BottomNavigationBarItem(icon: Icon(Icons.home), label: 'Home'),
          BottomNavigationBarItem(icon: Icon(Icons.person), label: 'User'),
        ],
      ),
    );
  }
}

class HomeScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return BaseWidget(title: 'Home', buttonText: 'Go to User', to: '/user');
  }
}

class UserScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return BaseWidget(title: 'User', buttonText: 'Go to Settings', to: '/settings');
  }
}

class SettingsScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return BaseWidget(title: 'Settings', buttonText: 'Go to User', to: '/user');
  }
}
```

</details>

**Comment by marcguilera** [comment]
Thanks! This worked.
